src/precice/impl/ParticipantImpl.cpp

```diff
@@ -1942,18 +1946,39 @@ ParticipantImpl::MappedSamples ParticipantImpl::mappedSamples() const
 
 // Reinitialization
 
-int ParticipantImpl::getTotalMeshChanges() const
+ParticipantImpl::MeshChanges ParticipantImpl::getTotalMeshChanges() const
 {
   PRECICE_TRACE();
   PRECICE_ASSERT(_allowsRemeshing);
   Event e("remesh.exchangeLocalMeshChanges", profiling::Synchronize);
-  int   localMeshesChanges = _meshLock.countUnlocked();
-  PRECICE_DEBUG("Mesh changes of rank: {}", localMeshesChanges);
 
-  int totalMeshesChanges = 0;
-  utils::IntraComm::allreduceSum(localMeshesChanges, totalMeshesChanges);
-  PRECICE_DEBUG("Mesh changes of participant: {}", totalMeshesChanges);
-  return totalMeshesChanges;
+  // Gather local changes
+  std::vector<double> localMeshChanges;
+  for (auto context : _accessor->usedMeshContexts()) {
+    localMeshChanges.push_back(_meshLock.check(context->mesh->getName()) ? 0.0 : 1.0);
+  }
+  PRECICE_DEBUG("Mesh changes of rank: {}", localMeshChanges);
+
+  // TODO implement int version of allreduceSum
+  std::vector<double> totalMeshChanges(localMeshChanges.size(), 0.0);
+  utils::IntraComm::allreduceSum(localMeshChanges, totalMeshChanges);
+
+  // Convert the doubles to int
+  MeshChanges totalMeshChangesI(totalMeshChanges.begin(), totalMeshChanges.end());
+  PRECICE_DEBUG("Mesh changes of participant: {}", totalMeshChangesI);
+  return totalMeshChangesI;
+}
+
+void ParticipantImpl::clearStamplesOfChangedMeshes(MeshChanges totalMeshChanges)
+{
+  auto meshContexts = _accessor->usedMeshContexts();
+  for (std::size_t i = 0; i < totalMeshChanges.size(); ++i) {
+    if (totalMeshChanges[i] > 0.0) {
+      for (auto d : meshContexts[i]->mesh->data()) {
+        d->timeStepsStorage().clear();
+      }
```

Comment:
Do this inside `Mesh` for better encapsulation? E.g. call `mesh->clearTimeStepsStorages()` and implement

```
void Mesh::clearTimeStepsStorages() {
  for (auto d : _data) {
    d->timeStepsStorage().clear();
  }
}
```

Note that there is also a `Mesh::clear()` doing something similar.

Reply:
Done